rfcs/0024-python-scope-package-set.md

```diff
@@ -0,0 +1,125 @@
+- `calibre` is a program for managing e-books that happens to be written in Python. The package is therefore considered an *application*
+- `pytest` is a test runner. While a program, it can only be used in conjunction with the Python environment where it needs to import the modules that are to be tested. It is therefore considered a *library*
+
+## Scope of the package set
+
+The purpose of the package set is to provide a set of compatible libraries that
+can be used for development purposes and as dependencies by Python applications.
+Applications are not considered part of the package set and their expressions
+shall thus also be placed elsewhere.
+
+For each NixOS release, a set of compatible libraries shall be formed
+that functions on all supported platforms, interpreter versions and
+implementations.
+
+That in effect means only one version of a package is permitted as having
+multiple versions in the set may lead to collisions. Exceptions can be made,
+like in the case of build or test-time dependencies like the `pytest` test runner.
+
+## Contributing and maintaining expressions
+
+Maintenance of the expressions takes effort. All expressions shall therefore
+have a maintainer. Packages that have no maintainer may be removed from the
+package set or marked as broken in case they seem to be unused, untouched in an
+extended period of time, or in whatever way cause additional work for other
+maintainers or package set maintainers. Guidelines for contributing expressions
+to the package set shall be made available in the Nixpkgs manual.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+-
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Use expression generators for package set
+
+Instead of manually creating a curated set of packages, an expression generator
+such as `pypi2nix` could be used. Using an expression generator could reduce the
+amount of work needed for writing and updating expressions. The `pypi2nix` tool
+uses `pip` to resolve dependencies.
+
+This is not recommended because:
+- non-Python dependencies are not specified upstream
+- excessive pinning of dependencies can lead to an unsatisfiable set
+- curated sets need to be generated for multiple combinations of platforms,
+interpreter versions and implementations. That requires being able to check out
+the upstream database (PyPI) at a certain state. While technically possible, no
+solutions are yet available.
+
+## Limit scope to application dependencies
+
+The scope of the package set could be limited to only providing dependencies for
+*applications*. Developers could use 3rd-party expression generators like
+`pypi2nix` to create expressions for development purposes. This is not
+recommended because:
+- applications can vary significantly in their dependencies, requiring different
+versions of packages, negating the use of a shared package set
+- of the reasons against using expression generators.
```

Comment:
It seems like a good idea given the current state of tooling. Hopefully we'll have something better than pypi2nix, maybe something more interactive, to build larger snapshots of packages working together.

Reply:
I would like pypi2nix to eventually be able to support cascading of package sets (that are defined in Nix). Basically, have a main package set, and then secondary sets that build upon and extend the first one, without repeating its values. That would be useful for our *applications*.